src/lib/formats/fs_hplif.cpp

```diff
@@ -0,0 +1,496 @@
+	hplif_dirent hdr = dirent.value();
+	std::vector<u8> result;
+	result.insert(result.end(),
+		      reinterpret_cast<const u8 *>(&hdr),
+		      reinterpret_cast<const u8 *>(&hdr) + 32);
+	result.reserve(sector_count * 256 + 32);
+	block_iterator iter(*this, big_endianize_int32(dirent->m_starting_sector), sector_count);
+	while (iter.next())
+		result.insert(result.end(), (const u8 *)iter.data(), (const u8 *)iter.data() + 256);
+
+	return std::make_pair(ERR_OK, std::move(result));
+}
+
+
+//-------------------------------------------------
+//  impl::read_sector
+//-------------------------------------------------
+
+fsblk_t::block_t impl::read_sector(u32 sector) const
+{
+	return m_blockdev.get(sector);
+}
+
+
+//-------------------------------------------------
+//  impl::dirent_from_path
+//-------------------------------------------------
+
+std::optional<impl::hplif_dirent> impl::dirent_from_path(const std::vector<std::string> &path) const
+{
+	if (path.size() != 1)
+		return { };
+	std::string_view path_part = path[0];
+
+	std::optional<hplif_dirent> result;
+	auto callback = [&result, path_part](const hplif_dirent &dirent)
+	{
+		bool found = strtrimright_hplif(dirent.m_file_name) == path_part;
+		if (found)
+			result = dirent;
+		return found;
+	};
+	iterate_directory_entries(callback);
+	return result;
+}
+
+
+//-------------------------------------------------
+//  impl::iterate_directory_entries
+//-------------------------------------------------
+
+void impl::iterate_directory_entries(const std::function<bool(const hplif_dirent &dirent)> &callback) const
+{
+	fsblk_t::block_t block = m_blockdev.get(0);
+	block_iterator iter(*this, block.r32b(8), block.r32b(16));
+
+	if (block.r16b(0) != 0x8000)
+	{
+		return;
+	}
+
+	while (iter.next())
+	{
+		for (const hplif_dirent &ent : iter.dirent_data())
+		{
+			if (ent.m_file_type == 0xffff)
+			{
+				return;
+			}
+
+			if (ent.m_file_type != 0x0000)
+			{
+				if (callback(ent))
+					return;
+			}
+		}
+	}
+}
+
+//-------------------------------------------------
+//  impl::decode_datetime
+//-------------------------------------------------
+
+util::arbitrary_datetime impl::decode_datetime(const hplif_time *time) const
+{
+	util::arbitrary_datetime result;
+	memset(&result, 0, sizeof(result));
+
+	result.year         = bcd_2_dec(time->year) + 1900;
+	result.month        = bcd_2_dec(time->month);
+	result.day_of_month = bcd_2_dec(time->day);
+	result.hour         = bcd_2_dec(time->hour);
+	result.minute       = bcd_2_dec(time->minute);
+	result.second       = bcd_2_dec(time->second);
+	return result;
+}
+
+//-------------------------------------------------
+//  impl::metadata_from_dirent
+//-------------------------------------------------
+
+meta_data impl::metadata_from_dirent(const hplif_dirent &dirent) const
+{
+	std::string file_type;
+
+	file_type = util::string_format("0x%04X", (int)big_endianize_int16(dirent.m_file_type));
```

Comment:
You may as well do this on a single line rather than a separate declaration and assignment.  Also, `string_format` doesn’t need the cast-to-`int` – it’s type-safe and will do the right thing.

Reply:
Forgot to remote that cast. I agree that it isn't neccessary.